From ec1a9b524cbbddd16f2c730c35f54b3a05f4c9a0 Mon Sep 17 00:00:00 2001
From: Harish Mahendrakar <harish.mahendrakar@ittiam.com>
Date: Fri, 3 Aug 2018 17:05:48 -0700
Subject: [PATCH 26/80] SoftXaac: Handle error cases when decoder fails to
 initialize

When decoder fails to initialize, it doesn't consume any bytes in
subsequent process call. Avoid this by returning an error when decoder
is not initialized before first process call.

Bug: 112194305
Test: cts-tradefed run commandAndExit cts-dev -m CtsSecurityTestCases\
 -t android.security.cts.StagefrightTest#testStagefright_c_2016_2428

Change-Id: I38ef56e8ff544567c3219c03dc6133c1ef98f3b3
---
 media/libstagefright/codecs/xaacdec/SoftXAAC.cpp | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/media/libstagefright/codecs/xaacdec/SoftXAAC.cpp b/media/libstagefright/codecs/xaacdec/SoftXAAC.cpp
index f173e0fba..06b15b3e5 100644
--- a/media/libstagefright/codecs/xaacdec/SoftXAAC.cpp
+++ b/media/libstagefright/codecs/xaacdec/SoftXAAC.cpp
@@ -689,7 +689,6 @@ void SoftXAAC::onQueueFilled(OMX_U32 /* portIndex */) {
                     notify(OMX_EventError, OMX_ErrorUndefined, err_code, NULL);
                     return;
                 }
-                mIsCodecConfigFlushRequired = true;
             }
 
             if (!mSampFreq || !mNumChannels) {
@@ -713,10 +712,14 @@ void SoftXAAC::onQueueFilled(OMX_U32 /* portIndex */) {
             signed int bytesConsumed = 0;
             int errorCode = 0;
             if (mIsCodecInitialized) {
+                mIsCodecConfigFlushRequired = true;
                 errorCode =
                     decodeXAACStream(inBuffer, inBufferLength, &bytesConsumed, &numOutBytes);
-            } else {
+            } else if (!mIsCodecConfigFlushRequired) {
                 ALOGW("Assumption that first frame after header initializes decoder failed!");
+                mSignalledError = true;
+                notify(OMX_EventError, OMX_ErrorUndefined, -1, NULL);
+                return;
             }
             inHeader->nFilledLen -= bytesConsumed;
             inHeader->nOffset += bytesConsumed;
-- 
2.17.1

